terraform/files/template/cluster-template.yaml

```diff
@@ -11,268 +51,283 @@ spec:
-          #
-          # Script exits without any defragmentation/backup/trim action if:
-          #  - It is executed on non leader etcd member
-          #  - It is executed on etcd cluster with some unhealthy member
-          #  - It is executed on single member etcd cluster
-          # Conditions above could be skipped and the script execution could be forced by the optional arguments:
-          #  - `--force-single`
-          #  - `--force-unhealthy`
-          #  - `--force-nonleader`
-          #
-          # The defragmentation on the etcd cluster is executed as follows:
-          #  - Defragment the non leader etcd members first
-          #  - Change the leadership to the randomly selected and defragmentation completed etcd member
-          #  - Defragment the local (ex-leader) etcd member
-          # Script then backup & trim local (ex-leader) etcd member
-          #
-          # Usage: etcd-defrag.sh [--force-single] [--force-unhealthy] [--force-nonleader]
-
-          export LOG_DIR=/var/log
-          export ETCDCTL_API=3
-          ETCDCTL="etcdctl --cert /etc/kubernetes/pki/etcd/peer.crt --key /etc/kubernetes/pki/etcd/peer.key --cacert /etc/kubernetes/pki/etcd/ca.crt"
+  template:
+    spec:
+#      replicas: ${CONTROL_PLANE_MACHINE_COUNT}
+#      machineTemplate:
+#        infrastructureRef:
+#          kind: OpenStackMachineTemplate
+#          # apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+#          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha7
+#          name: "${PREFIX}-${CLUSTER_NAME}-control-plane-${CONTROL_PLANE_MACHINE_GEN}"
+      kubeadmConfigSpec:
+        initConfiguration:
+          nodeRegistration:
+            name: '{{ local_hostname }}'
+            kubeletExtraArgs:
+              cloud-provider: external
+              provider-id: openstack:///'{{ instance_id }}'
+        clusterConfiguration:
+          imageRepository: registry.k8s.io
+          dns:
+            imageRepository: registry.k8s.io/coredns
+          apiServer:
+            extraArgs:
+              cloud-provider: external
+          controllerManager:
+            extraArgs:
+              cloud-provider: external
+          etcd:
+            local:
+              dataDir: /var/lib/etcd
+              extraArgs:
+                heartbeat-interval: "250"
+                election-timeout: "2500"
+                snapshot-count: "6400"
+                auto-compaction-mode: periodic
+                auto-compaction-retention: 8h
+        joinConfiguration:
+          nodeRegistration:
+            name: '{{ local_hostname }}'
+            kubeletExtraArgs:
+              cloud-provider: external
+              provider-id: openstack:///'{{ instance_id }}'
+        files:
+          - path: /root/etcd-defrag.sh
+            owner: "root:root"
+            permissions: "0755"
+            content: |
+              #!/bin/bash
```

Comment:
There seem to be a lot of unnecessary (whitespace ...) changes to the scripts.
I'd prefer them not to be included, as it complicates the review process and makes searching for issues with git annotate / git blame much harder.
If you think that this is valuable cleanup, please submit it in a separate PR.

Reply:
Those are not whitespaces, but I moved the whole .spec of to .spec.template.spec so 4 spaces to the right. It is because we need `KubeadmControlPlaneTemplate` instead of `KubeadmControlPlane`.